### PR TITLE
support for pytrees in projection_halfspace and projection_hyperplane

### PR DESCRIPTION
**Pull request by gowerrobert**
1. re-implemented projection_halfspace and projection_hyperplane so they work with pytrees
2. added tests to reflect support
3. also added a test for projection_box,

### CODE REVIEW

File: jaxopt/_src/projection.py [hunk 1]
```diff
@@ -219,14 +221,15 @@ def projection_hyperplane(x: jnp.ndarray, hyperparams: Tuple) -> jnp.ndarray:
 
   Args:
     x: pytree to project.
-    hyperparams: tuple ``hyperparams = (a, b)``, where ``a`` is a vector and
-      ``b`` is a scalar.
+    hyperparams: tuple ``hyperparams = (a, b)``, where ``a`` is a pytree with 
+                 the same structure as x ``b`` is a scalar.
```
Comment (mblondel) [inline jaxopt/_src/projection.py:225]:
```suggestion
                 the same structure as ``x`` and ``b`` is a scalar.
```

File: jaxopt/_src/projection.py [hunk 2]
```diff
@@ -219,14 +221,15 @@ def projection_hyperplane(x: jnp.ndarray, hyperparams: Tuple) -> jnp.ndarray:
...
 
   Returns:
     output array, with the same shape as ``x`.
   """
   a, b = hyperparams
-  return x - (jnp.dot(a, x) - b) / jnp.dot(a, a) * a
+  scale = (tree_util.tree_vdot(a,x) -b)/tree_util.tree_vdot(a,a)
```
Comment (mblondel) [inline jaxopt/_src/projection.py:231]:
```suggestion
  scale = (tree_util.tree_vdot(a,x) -b) / tree_util.tree_vdot(a,a)
```

File: jaxopt/_src/projection.py [hunk 3]
```diff
@@ -239,14 +242,15 @@ def projection_halfspace(x: jnp.ndarray, hyperparams: Tuple) -> jnp.ndarray:
 
   Args:
     x: pytree to project.
-    hyperparams: tuple ``hyperparams = (a, b)``, where ``a`` is a vector and
-      ``b`` is a scalar.
+    hyperparams: tuple ``hyperparams = (a, b)``, where ``a`` is a pytree with 
+                 the same structure as x and ``b`` is a scalar.
```
Comment (mblondel) [inline jaxopt/_src/projection.py:246]:
```suggestion
                 the same structure as ``x`` and ``b`` is a scalar.
```

File: jaxopt/_src/projection.py [hunk 4]
```diff
@@ -239,14 +242,15 @@ def projection_halfspace(x: jnp.ndarray, hyperparams: Tuple) -> jnp.ndarray:
...
 
   Returns:
     output array, with same shape as ``x``.
   """
   a, b = hyperparams
-  return x - jax.nn.relu(jnp.dot(a, x) - b) / jnp.dot(a, a) * a
+  scale = jax.nn.relu(tree_util.tree_vdot(a,x) -b)/tree_util.tree_vdot(a,a)
```
Comment (mblondel) [inline jaxopt/_src/projection.py:252]:
```suggestion
  scale = jax.nn.relu(tree_util.tree_vdot(a, x) - b) / tree_util.tree_vdot(a, a)
```

File: jaxopt/_src/projection.py [hunk 5]
```diff
@@ -42,6 +42,8 @@ def projection_non_negative(x: Any, hyperparams=None) -> Any:
   del hyperparams  # Not used.
   return tree_util.tree_map(jax.nn.relu, x)
 
+def _clip_safe(x, a, b):
```
Comment (mblondel) [inline jaxopt/_src/projection.py:45]:
```suggestion
def _clip_safe(x, lower, upper):
```

File: tests/projection_test.py [hunk 6]
```diff
@@ -45,6 +49,14 @@ def test_projection_box(self):
     self.assertArraysEqual(projection.projection_box((x, x), params),
                            (expected, expected))
 
+    # lower and upper are pytrees. Note: Does not work when U and L are scalars!
```
Comment (mblondel) [inline tests/projection_test.py:52]: I think the following could potentially work:
```python
if isinstance(lower_bound, float):
  # If lower_bound is a scalar, create a pytree with the same structure as `x`
  lower_bound = tree_util.tree_zeros_like(x) + lower_bound
```
but let's do it in a separate PR.

File: tests/projection_test.py [hunk 7]
```diff
@@ -222,6 +234,21 @@ def test_projection_hyperplane(self):
...
+    self.assertAllClose(expected_p, p)
+
+  def test_projection_halfspace(self):
+    # Testing that for b large enough, halfplane projection of x is x.
+    rng = onp.random.RandomState(0)
+    tree_x = (rng.randn(1), {"k1": rng.randn(1), "k2": (rng.randn(1), rng.randn(1))}, rng.randn(1))
+    tree_a = (rng.randn(1), {"k1": rng.randn(1), "k2": (rng.randn(1), rng.randn(1))}, rng.randn(1))
+    b = 10000
+    p = projection.projection_halfspace(tree_x, (tree_a, b))
+    self.assertAllClose(tree_x, p)
```
Comment (mblondel) [inline tests/projection_test.py:250]: Thanks for creating this test! I hadn't realized that we were missing it. 

Could you also check the case when `b` is small? A simple assertion is to just check that the solution is feasible. A more advanced assertion would be to check that the optimality conditions are satisfied.

**Review by mblondel — CHANGES_REQUESTED** [review]
Thanks a lot Robert!

I left some comments

**Comment by gowerrobert** [comment]
Thanks for the suggestions and comments @mblondel  ! I've incorporated them all (except the further fix for box projections). Shall I squash the commits into one?

**Comment by mblondel** [comment]
Yes please :)

Le mer. 29 sept. 2021 à 17:50, Robert M. Gower ***@***.***> a
écrit :

**Comment by gowerrobert** [comment]
Ok, I think I managed to squash the commits (excluding a merge commit that came for free)

**Comment by mblondel** [comment]
Did you force push? There are currently 8 commits.

**Comment by gowerrobert** [comment]
nope, I hadn't forced the push. Now it's right (I hope)

**Pull request merged**